adapters/smarthub/smarthub.go

```diff
@@ -0,0 +1,187 @@
+package smarthub
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"text/template"
+
+	"github.com/mxmCherry/openrtb/v15/openrtb2"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/macros"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+const (
+	ADAPTER_VER = "1.0.0"
+)
+
+type adapter struct {
+	endpoint template.Template
+}
+
+type bidExt struct {
+	MediaType string `json:"mediaType"`
+}
+
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter) (adapters.Bidder, error) {
+	template, err := template.New("endpointTemplate").Parse(config.Endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse endpoint url template: %v", err)
+	}
+
+	bidder := &adapter{
+		endpoint: *template,
+	}
+
+	return bidder, nil
+}
+
+func (a *adapter) getImpressionExt(imp *openrtb2.Imp) (*openrtb_ext.ExtSmartHub, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: "Bidder extension not provided or can't be unmarshalled",
+		}
+	}
+
+	var smarthubExt openrtb_ext.ExtSmartHub
+	if err := json.Unmarshal(bidderExt.Bidder, &smarthubExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: "Error while unmarshaling bidder extension",
+		}
+	}
+
+	return &smarthubExt, nil
+}
+
+func (a *adapter) buildEndpointURL(params *openrtb_ext.ExtSmartHub) (string, error) {
+	endpointParams := macros.EndpointTemplateParams{
+		Host:      params.Host,
+		AccountID: params.Seat,
+		SourceId:  params.Token,
+	}
+	return macros.ResolveMacros(a.endpoint, endpointParams)
+}
+
+func (a *adapter) MakeRequests(
+	openRTBRequest *openrtb2.BidRequest,
+	reqInfo *adapters.ExtraRequestInfo,
+) (
+	requestsToBidder []*adapters.RequestData,
+	errs []error,
+) {
+	if len(openRTBRequest.Imp) == 0 {
+		return nil, []error{&errortypes.BadInput{Message: "Missing Imp object"}}
```

Comment:
Nitpick: Prebid server core filters out requests with empty `[]Imp` arrays so there's no need for this check. I realize this isn't a lot overhead so you could leave it or remove it if you want.

Reply:
Removed this